// @esri/hub-common/dist/esm/groups/add-users-workflow/workflow-sections/invite-users.js
import { inviteGroupUsers, } from "@esri/arcgis-rest-portal";
/**
 *
 * Attempts to invite users to a group
 *
 * @param {string} id ID of the group the users will be invited to
 * @param {object[]} users
 * @param {object} authentication
 * @param {number} expiration How long the invite will be active (in minutes)
 * @param {string} role What role should they be added as. Defaults to group member
 *
 * @returns {object|null} Result of the transaction (null if no users are passed in)
 */
export function inviteUsers(id, users, authentication, expiration, // default to 2 week expiration TODO: is this actually 2 weeks?
role // default to group member, but allow for team_admin as well
) {
    if (expiration === void 0) { expiration = 20160; }
    if (role === void 0) { role = "group_member"; }
    var response = Promise.resolve(null);
    if (users.length) {
        var args = {
            id: id,
            users: users.map(function (u) { return u.username; }),
            authentication: authentication,
            role: role,
            expiration: expiration,
        };
        response = inviteGroupUsers(args);
    }
    return response;
}
