// @esri/hub-search/dist/node/content/helpers/convert-portal-response.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertPortalResponse = void 0;
const arcgis_rest_portal_1 = require("@esri/arcgis-rest-portal");
const hub_common_1 = require("@esri/hub-common");
/**
 * Converts the response format returned by the Portal API to a common format
 * @param request - the ISearchOptions instance used to invoke the request
 * @param response - the ISearchResult JSON returned by the Portal API
 */
function convertPortalResponse(request, response) {
    const results = response.results.map(hub_common_1.itemToContent);
    const count = response.num;
    const total = response.total;
    const hasNext = response.nextStart > -1;
    const query = response.query;
    const aggregations = response.aggregations
        ? mapAggregations(response.aggregations)
        : undefined;
    const next = getNextFunction(request, response.nextStart, response.total);
    return {
        results,
        count,
        total,
        hasNext,
        query,
        aggregations,
        next,
    };
}
exports.convertPortalResponse = convertPortalResponse;
function mapAggregations(aggregations) {
    return Object.keys(aggregations).reduce((contentAggs, aggType) => {
        // Built in as a safety if Portal returns unsupported aggregations
        /* istanbul ignore else */
        if (aggType.toLowerCase() === "counts") {
            contentAggs.counts = mapCountAggregations(aggregations[aggType]);
        }
        return contentAggs;
    }, {});
}
function mapCountAggregations(countAggs) {
    return countAggs.map((agg) => {
        const mappedAggs = agg.fieldValues.map((aggValue) => ({
            label: aggValue.value,
            value: aggValue.count,
        }));
        return {
            fieldName: agg.fieldName,
            aggregations: mappedAggs,
        };
    });
}
function getNextFunction(request, nextStart, total) {
    const clonedRequest = hub_common_1.cloneObject(request);
    // Authentication not properly cloned
    clonedRequest.authentication = request.authentication;
    clonedRequest.start = nextStart > -1 ? nextStart : total + 1;
    return (authentication) => {
        if (authentication) {
            clonedRequest.authentication = authentication;
        }
        return arcgis_rest_portal_1.searchItems(clonedRequest).then((response) => {
            return convertPortalResponse(clonedRequest, response);
        });
    };
}
