// @esri/hub-common/dist/esm/auth.js
/* Copyright (c) 2018 Environmental Systems Research Institute, Inc.
 * Apache-2.0 */
import { request } from "@esri/arcgis-rest-request";
import { UserSession } from "@esri/arcgis-rest-auth";
/**
 * A thin wrapper around [`UserSession.completeOAuth2()`](https://esri.github.io/arcgis-rest-js/api/auth/UserSession/#completeOAuth2) that sets search tags and other relevant metadata for newly created community users.
 */
/* istanbul ignore next */
export function completeOAuth2(options, win) {
    if (win === void 0) { win = window; }
    var match = win.location.href.match(/access_token=(.+)&expires_in=.+&username=([^&]+)/);
    var token = match[1];
    var user = decodeURIComponent(match[2]);
    var baseUrl = "https://www.arcgis.com/sharing/rest/community/users/" + user;
    return request(baseUrl, {
        params: { token: token },
        httpMethod: "GET"
    }).then(function (response) {
        if (Date.now() - response.created < 5000) {
            return request(baseUrl + "/update", {
                params: {
                    token: token,
                    tags: ["hubRole:participant", "org:" + response.orgId],
                    access: "public"
                }
            }).then(function () {
                return UserSession.completeOAuth2(options);
            });
        }
        else {
            return UserSession.completeOAuth2(options);
        }
    });
}
