// @esri/hub-common/dist/esm/categories.js
import { collections } from "./collections";
var app = collections.app, dataset = collections.dataset, document = collections.document, event = collections.event, feedback = collections.feedback, initiative = collections.initiative, map = collections.map, other = collections.other, site = collections.site;
var downloadableTypes = [
    "360 VR Experience",
    "Application",
    "CityEngine Web Scene",
    "Code Sample",
    "CSV Collection",
    "CSV",
    "CAD Drawing",
    "Desktop Application",
    "Desktop Application Template",
    "Desktop Style",
    "File Geodatabase",
    "GeoJson",
    "Geoprocessing Package",
    "Geoprocessing Sample",
    "Image",
    "iWork Keynote",
    "iWork Numbers",
    "KML Collection",
    "KML",
    "Layer",
    "Layer File",
    "Layer Package",
    "Layout",
    "Locator Package",
    "Map Package",
    "Map Service Definition",
    "Map Template",
    "Microsoft Excel",
    "Microsoft Powerpoint",
    "Microsoft Visio",
    "Microsoft Word",
    "Notebook",
    "Operations Dashboard Add In",
    "PDF",
    "Pro Map",
    "Project Package",
    "Project Template",
    "Raster function template",
    "Rule Package",
    "Service Definition",
    "Shapefile",
    "Vector Tile Package",
    "Workflow Manager Package",
];
var downloadableTypeKeywords = ["Data"];
var apiTypes = ["Feature Service", "Map Service", "Image Service"];
// TODO: remove this at next breaking version
// we're just keeping this for backwards compatibility
export var categories = {
    app: app.concat(feedback),
    dataset: dataset,
    document: document,
    event: event,
    initiative: initiative,
    map: map,
    other: other,
    site: site,
    downloadableTypes: downloadableTypes,
    downloadableTypeKeywords: downloadableTypeKeywords,
    apiTypes: apiTypes,
};
// TODO: move this function and supporting arrays to another module
/**
 * Is the item type downloadable in the Hub app
 * @param item ArcGIS item with type and type keywords
 */
export function isDownloadable(item) {
    return (downloadableTypes.indexOf(item.type) !== -1 ||
        (item.typeKeywords &&
            downloadableTypeKeywords.some(function (downloadableTypeKeyword) {
                return item.typeKeywords.some(function (typeKeyword) { return typeKeyword === downloadableTypeKeyword; });
            })));
}
