// @esri/hub-common/dist/node/sites/themes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getOrgDefaultTheme = exports.DEFAULT_THEME = void 0;
const __1 = require("..");
/**
 * Default Site Theme
 */
exports.DEFAULT_THEME = {
    header: {
        background: "#fff",
        text: "#4c4c4c",
    },
    body: {
        background: "#fff",
        text: "#4c4c4c",
        link: "#0079c1",
    },
    button: {
        background: "#0079c1",
        text: "#fff",
    },
    logo: {
        small: "",
    },
    fonts: {
        base: {
            url: "",
            family: "Avenir Next",
        },
        heading: {
            url: "",
            family: "Avenir Next",
        },
    },
};
/**
 * Return the default theme, extended with values from the Org's shared theme
 * @param {Object} portalSelf Org's Portal object
 */
function getOrgDefaultTheme(portalSelf) {
    let defaultTheme = __1.cloneObject(exports.DEFAULT_THEME);
    let sharedTheme = __1.getProp(portalSelf, "portalProperties.sharedTheme");
    if (sharedTheme) {
        sharedTheme = __1.removeEmptyProps(sharedTheme);
        defaultTheme = __1.extend(defaultTheme, sharedTheme);
    }
    return defaultTheme;
}
exports.getOrgDefaultTheme = getOrgDefaultTheme;
