// @esri/hub-common/dist/node/sites/_internal/_ensure-telemetry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._ensureTelemetry = void 0;
const __1 = require("../..");
/**
 * Add telemetry config object
 * @private
 * @param {object} model Site Model
 * @returns {object}
 */
function _ensureTelemetry(model) {
    if (__1.getProp(model, "item.properties.schemaVersion") >= 1.4)
        return model;
    const clone = __1.cloneObject(model);
    const gacode = __1.getProp(clone, "data.values.gacode");
    clone.data.values.telemetry = {
        consentNotice: {
            isTheme: true,
            consentText: "",
            policyURL: "",
        },
        customAnalytics: {
            ga: {
                customerTracker: {
                    enabled: Boolean(gacode),
                    id: gacode,
                },
            },
        },
    };
    __1.deleteProp(clone, "data.values.gacode");
    __1.setProp("item.properties.schemaVersion", 1.4, clone);
    return clone;
}
exports._ensureTelemetry = _ensureTelemetry;
