// @esri/hub-common/dist/esm/utils/logger.js
/* tslint:disable no-console */
import { __spreadArrays } from "tslib";
/**
 * Enum for Logger Levels
 */
export var Level;
(function (Level) {
    Level[Level["all"] = 0] = "all";
    Level[Level["debug"] = 1] = "debug";
    Level[Level["info"] = 2] = "info";
    Level[Level["warn"] = 3] = "warn";
    Level[Level["error"] = 4] = "error";
    Level[Level["off"] = 5] = "off";
})(Level || (Level = {}));
/**
 * ```js
 * import { Logger, Level } from '@esri/hub-common'
 * ```
 * Functions share the console interface
 * ```js
 * Logger.log('My Message');
 * Logger.warn('Watch out!', { threat: 'Charizard' });
 * // etc, etc
 * ```
 * Available logging levels are specified in the Level enum. The hierarchy is as follows:
 * ```
 * off > error > warn > info > debug > all
 * ```
 * Logger only sends messages whose level is greater than or equal to the global log level
 * ```js
 * // Global level is 'warn'
 * Logger.info('This message won't log');
 * Logger.error('But this one will!');
 * ```
 * Logger's default level is 'off', so set desired level before use
 * ```js
 * Logger.setLogLevel(Level.all);
 * ```
 */
var Logger = /** @class */ (function () {
    function Logger() {
    }
    /**
     * Sets the global log level
     * @param {Level} level
     */
    Logger.setLogLevel = function (level) {
        this.logLevel = level;
    };
    /**
     * Logs to debug if level is enabled
     * @param {string} message
     * @param {...*} objects additional objects to log (optional rest parameter)
     */
    Logger.log = function (message) {
        var objects = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            objects[_i - 1] = arguments[_i];
        }
        if (this.isLevelPermitted(Level.debug)) {
            console.log.apply(console, __spreadArrays([message], objects));
        }
    };
    /**
     * Logs to debug if level is enabled
     * @param {string} message
     * @param {...*} objects additional objects to log (optional rest parameter)
     */
    Logger.debug = function (message) {
        var objects = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            objects[_i - 1] = arguments[_i];
        }
        if (this.isLevelPermitted(Level.debug)) {
            console.debug.apply(console, __spreadArrays([message], objects));
        }
    };
    /**
     * Logs to info if level is enabled
     * @param {string} message
     * @param {...*} objects additional objects to log (optional rest parameter)
     */
    Logger.info = function (message) {
        var objects = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            objects[_i - 1] = arguments[_i];
        }
        if (this.isLevelPermitted(Level.info)) {
            console.info.apply(console, __spreadArrays([message], objects));
        }
    };
    /**
     * Logs to warn if level is enabled
     * @param {string} message
     * @param {...*} objects additional objects to log (optional rest parameter)
     */
    Logger.warn = function (message) {
        var objects = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            objects[_i - 1] = arguments[_i];
        }
        if (this.isLevelPermitted(Level.warn)) {
            console.warn.apply(console, __spreadArrays([message], objects));
        }
    };
    /**
     * Logs to error if level is enabled
     * @param {string} message
     * @param {...*} objects additional objects to log (optional rest parameter)
     */
    Logger.error = function (message) {
        var objects = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            objects[_i - 1] = arguments[_i];
        }
        if (this.isLevelPermitted(Level.error)) {
            console.error.apply(console, __spreadArrays([message], objects));
        }
    };
    Logger.isLevelPermitted = function (level) {
        return this.logLevel <= level;
    };
    Logger.logLevel = Level.off;
    return Logger;
}());
export { Logger };
