// @esri/hub-initiatives/dist/esm/util.js
import { addItemResource } from "@esri/arcgis-rest-portal";
import { getPortalApiUrl } from "@esri/hub-common";
/**
 *  Copy an set of image resources from one item to another
 *
 * @export
 * @param {string} sourceItemId
 * @param {string} targetItemId
 * @param {string} owner
 * @param {[string]} assets
 * @param {IRequestOptions} requestOptions
 * @returns {Promise<boolean>}
 */
export function copyImageResources(sourceItemId, targetItemId, owner, assets, requestOptions) {
    var itemResourceUrl = getPortalApiUrl(requestOptions) + "/content/items/" + sourceItemId + "/resources";
    /* istanbul ignore next blob responses are difficult to make cross platform we will just have to trust the isomorphic fetch will do its job */
    return requestOptions.authentication
        .getToken(itemResourceUrl)
        .then(function (token) {
        var assetPromises = assets.map(function (assetName) {
            var sourceUrl = itemResourceUrl + "/" + assetName + "?token=" + token;
            return addImageAsResource(targetItemId, owner, assetName, sourceUrl, requestOptions);
        });
        // This is really more of a fire-and-forget thing, as the Portal API
        // adds these requests into a queue for processing
        return Promise.all(assetPromises);
    })
        .then(function () {
        return true;
    });
}
/**
 *  Copy an set of embedded resources to an item
 *
 * @export
 * @param {string} targetItemId destination item id
 * @param {string} owner destination item owner
 * @param {any[]} assets list of assets to copy
 * @param {IRequestOptions} requestOptions
 * @returns {Promise<boolean>}
 */
export function copyEmbeddedImageResources(targetItemId, owner, assets, requestOptions) {
    // need to move resources from embedded template into AGO
    var promises = assets.map(function (asset) {
        return addImageAsResource(targetItemId, owner, asset.name, asset.url, requestOptions)
            .then(function () {
            return true;
        })
            .catch(function () {
            return true; // swallow the error
        });
    });
    return Promise.all(promises).then(function () {
        return true;
    });
}
/**
 * Load an image from a url, and store it as a resource on an existing item
 *
 * @export
 * @param {string} itemId
 * @param {string} owner
 * @param {string} filename
 * @param {string} url
 * @param {IRequestOptions} requestOptions
 * @returns {Promise<boolean>}
 */
export function addImageAsResource(id, owner, name, url, requestOptions) {
    // We must use fetch directly here because AGSjs request determines how to parse
    // the response based on the `?f=<type>` param. But sending f=<anything-other-than-json>
    // to a resource end-point, throws an error. Rock, meet Hard-Place.
    var fetchOptions = {
        method: "GET",
        // ensures behavior mimics XMLHttpRequest. needed to support sending IWA cookies
        credentials: "same-origin"
    };
    return (
    // -------------------------------------------------
    // request(url, opts)
    // -------------------------------------------------
    fetch(url, fetchOptions)
        .then(function (x) {
        return x.blob();
    })
        .then(function (blob) {
        return addItemResource({
            id: id,
            owner: owner,
            name: name,
            resource: blob,
            authentication: requestOptions.authentication
        }).then(function (response) {
            return response.success;
        });
    }));
}
