// @esri/hub-search/dist/node/ago/format-item-collection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.agoFormatItemCollection = void 0;
const format_1 = require("./helpers/format");
// This function is responsible for formatting results from the AGO API into the JSONAPI collection format
function agoFormatItemCollection(searchResults, facets = {}, params) {
    const queryParams = Object.assign({}, params);
    return {
        data: searchResults.results.map(result => {
            return format_1.formatItem(result, queryParams.q);
        }),
        meta: {
            query: searchResults.query,
            queryParameters: queryParams,
            stats: {
                aggs: facets,
                count: searchResults.results.length,
                totalCount: searchResults.total
            },
            page: {
                start: searchResults.start,
                size: searchResults.num,
                nextStart: searchResults.nextStart
            }
        }
    };
}
exports.agoFormatItemCollection = agoFormatItemCollection;
