// @esri/hub-search/dist/esm/content/helpers/convert-portal-response.js
import { searchItems, } from "@esri/arcgis-rest-portal";
import { cloneObject, itemToContent } from "@esri/hub-common";
/**
 * Converts the response format returned by the Portal API to a common format
 * @param request - the ISearchOptions instance used to invoke the request
 * @param response - the ISearchResult JSON returned by the Portal API
 */
export function convertPortalResponse(request, response) {
    var results = response.results.map(itemToContent);
    var count = response.num;
    var total = response.total;
    var hasNext = response.nextStart > -1;
    var query = response.query;
    var aggregations = response.aggregations
        ? mapAggregations(response.aggregations)
        : undefined;
    var next = getNextFunction(request, response.nextStart, response.total);
    return {
        results: results,
        count: count,
        total: total,
        hasNext: hasNext,
        query: query,
        aggregations: aggregations,
        next: next,
    };
}
function mapAggregations(aggregations) {
    return Object.keys(aggregations).reduce(function (contentAggs, aggType) {
        // Built in as a safety if Portal returns unsupported aggregations
        /* istanbul ignore else */
        if (aggType.toLowerCase() === "counts") {
            contentAggs.counts = mapCountAggregations(aggregations[aggType]);
        }
        return contentAggs;
    }, {});
}
function mapCountAggregations(countAggs) {
    return countAggs.map(function (agg) {
        var mappedAggs = agg.fieldValues.map(function (aggValue) { return ({
            label: aggValue.value,
            value: aggValue.count,
        }); });
        return {
            fieldName: agg.fieldName,
            aggregations: mappedAggs,
        };
    });
}
function getNextFunction(request, nextStart, total) {
    var clonedRequest = cloneObject(request);
    // Authentication not properly cloned
    clonedRequest.authentication = request.authentication;
    clonedRequest.start = nextStart > -1 ? nextStart : total + 1;
    return function (authentication) {
        if (authentication) {
            clonedRequest.authentication = authentication;
        }
        return searchItems(clonedRequest).then(function (response) {
            return convertPortalResponse(clonedRequest, response);
        });
    };
}
