// @esri/hub-common/dist/es2017/projects/HubProjects.js
// Note - we separate these imports so we can cleanly spy on things in tests
import { createModel, fetchModelFromItem, getModel, updateModel, } from "../models";
import { constructSlug, getItemBySlug, getUniqueSlug, setSlugKeyword, } from "../items/slugs";
import { isGuid, cloneObject, mergeContentFilter, getItemThumbnailUrl, } from "..";
import { removeItem, getItem, } from "@esri/arcgis-rest-portal";
import { searchContentEntities } from "../search/_internal/searchContentEntities";
import { PropertyMapper } from "../core/_internal/PropertyMapper";
export const HUB_PROJECT_ITEM_TYPE = "Hub Project";
/**
 * Default values of a IHubProject
 */
const DEFAULT_PROJECT = {
    name: "No title provided",
    tags: [],
    typeKeywords: ["Hub Project"],
    status: "inactive",
};
/**
 * Default values for a new HubProject Model
 */
const DEFAULT_PROJECT_MODEL = {
    item: {
        type: HUB_PROJECT_ITEM_TYPE,
        title: "No Title Provided",
        description: "No Description Provided",
        snippet: "",
        tags: [],
        typeKeywords: ["Hub Project"],
        properties: {
            slug: "",
        },
    },
    data: {
        display: "about",
        timeline: {},
        status: "inactive",
        contacts: [],
        schemaVersion: 1,
    },
};
/**
 * Returns an Array of IPropertyMap objects
 * We could define these directly, but since the
 * properties of IHubProject map directly to properties
 * on item or data, it's slightly less verbose to
 * generate the structure.
 * @returns
 */
function getProjectPropertyMap() {
    const itemProps = [
        "created",
        "culture",
        "description",
        "extent",
        "id",
        "modified",
        "owner",
        "snippet",
        "tags",
        "typeKeywords",
        "url",
        "type",
    ];
    const dataProps = [
        "contacts",
        "display",
        "geometry",
        "headerImage",
        "layout",
        "location",
        "status",
    ];
    const map = [];
    itemProps.forEach((entry) => {
        map.push({ objectKey: entry, modelKey: `item.${entry}` });
    });
    dataProps.forEach((entry) => {
        map.push({ objectKey: entry, modelKey: `data.${entry}` });
    });
    // Deeper mappings
    map.push({
        objectKey: "slug",
        modelKey: "item.properties.slug",
    });
    map.push({
        objectKey: "orgUrlKey",
        modelKey: "item.properties.orgUrlKey",
    });
    map.push({
        objectKey: "name",
        modelKey: "item.title",
    });
    return map;
}
/**
 * Create a new Hub Project item
 *
 * Minimal properties are name and org
 *
 * @param project
 * @param requestOptions
 */
export async function createProject(partialProject, requestOptions) {
    // merge incoming with the default
    // this expansion solves the typing somehow
    const project = Object.assign(Object.assign({}, DEFAULT_PROJECT), partialProject);
    // Create a slug from the title if one is not passed in
    if (!project.slug) {
        project.slug = constructSlug(project.name, project.orgUrlKey);
    }
    // Ensure slug is  unique
    project.slug = await getUniqueSlug({ slug: project.slug }, requestOptions);
    // add slug to keywords
    project.typeKeywords = setSlugKeyword(project.typeKeywords, project.slug);
    // Map project object onto a default project Model
    const mapper = new PropertyMapper(getProjectPropertyMap());
    // create model from object, using the default model as a starting point
    let model = mapper.objectToModel(project, cloneObject(DEFAULT_PROJECT_MODEL));
    // create the item
    model = await createModel(model, requestOptions);
    // map the model back into a IHubProject
    const newProject = mapper.modelToObject(model, {});
    // and return it
    return newProject;
}
/**
 * Update a Hub Project
 * @param project
 * @param requestOptions
 */
export async function updateProject(project, requestOptions) {
    // verify that the slug is unique, excluding the current project
    project.slug = await getUniqueSlug({ slug: project.slug, existingId: project.id }, requestOptions);
    // get the backing item & data
    const model = await getModel(project.id, requestOptions);
    // create the PropertyMapper
    const mapper = new PropertyMapper(getProjectPropertyMap());
    // Note: Although we are fetching the model, and applying changes onto it,
    // we are not attempting to handle "concurrent edit" conflict resolution
    // but this is where we would apply that sort of logic
    const modelToUpdate = mapper.objectToModel(project, model);
    // update the backing item
    const updatedModel = await updateModel(modelToUpdate, requestOptions);
    // now map back into a project and return that
    const updatedProject = mapper.modelToObject(updatedModel, project);
    // the casting is needed because modelToObject returns a `Partial<T>`
    // where as this function returns a `T`
    return updatedProject;
}
/**
 * Get a Hub Project by id or slug
 * @param identifier item id or slug
 * @param requestOptions
 */
export function fetchProject(identifier, requestOptions) {
    let getPrms;
    if (isGuid(identifier)) {
        // get item by id
        getPrms = getItem(identifier, requestOptions);
    }
    else {
        getPrms = getItemBySlug(identifier, requestOptions);
    }
    return getPrms.then((item) => {
        if (!item)
            return null;
        return convertItemToProject(item, requestOptions);
    });
}
/**
 * Remove a Hub Project
 * @param id
 * @param requestOptions
 */
export async function destroyProject(id, requestOptions) {
    const ro = Object.assign(Object.assign({}, requestOptions), { id });
    await removeItem(ro);
    return;
}
/**
 * Search for Projects, and get IHubProject results
 *
 * Different from `searchContent` in that this returns the specific entity type
 *
 * @param filter
 * @param options
 * @returns
 */
export async function searchProjects(filter, options) {
    // Scope to Hub Projects
    const scopingFilter = {
        filterType: "content",
        type: {
            exact: ["Hub Project"],
        },
    };
    // merge filters
    const projectFilter = mergeContentFilter([scopingFilter, filter]);
    return searchContentEntities(projectFilter, convertItemToProject, options);
}
/**
 * Convert an Hub Project Item into a Hub Project, fetching any additional
 * information that may be required
 * @param item
 * @param auth
 * @returns
 */
export async function convertItemToProject(item, requestOptions) {
    const model = await fetchModelFromItem(item, requestOptions);
    const mapper = new PropertyMapper(getProjectPropertyMap());
    let token;
    if (requestOptions.authentication) {
        const session = requestOptions.authentication;
        token = session.token;
    }
    const prj = mapper.modelToObject(model, {});
    prj.thumbnailUrl = getItemThumbnailUrl(model.item, requestOptions, token);
    return prj;
}
